Edit pass and additions on best practices (#2)
garyillyes/cbcp: draft-illyes-cbcp.md

--- abstract

This document describes best pratices for web crawlers.

--- middle

# Introduction

Automatic clients (i.e. crawlers, bots) are used to access web resources, e.g., for indexing
for search but more recently also for new use cases related to Artificial Intelligence (AI)
such as training models. With an increase in crawling activity, it is particular important 
that automatic clients having an expected behavior and respect contraint of the ressources
being crawled. This includes documenting how to identify them and how their behavior
can be influenced. As such, crawler operators are requested
to conform to crawling best pratices in this document.

To further help website owners,

> it should in additon be considered to create together a central

Author: This was in your text already ;-) But I'm fine to remove it.

> it should in additon be considered to create together a central

Reviewer: My bad. What I mean is that we should figure out how this would be feasible because I can't think of a way either site/service owners or crawler operators could reasonably maintain such a registry.

> it should in additon be considered to create together a central

Author: ```suggestion
it should in additon be considered to create a central
```

> it should in additon be considered to create together a central

Author: Do my two small proposed word changes work for you?

> it should in additon be considered to create together a central

Reviewer: yup

> website where website owners can look up well behaving crawlers.

Author: ```suggestion
registry where website owners can look up well behaving crawlers.
```

Note that while self declared research crawlers, including privacy and malware discovery
crawlers, and contractual crawlers are welcome to add themselves to adopt these practices,
due to the nature of the relationship with sites they may exempt themselves from any of
the Crawler Code of Conduct policies with a rationale.

# Recommended Best Practices

The following best practices are should be followed and are already
applied by a vast majority of large scale crawlers on the Internet:

1. Crawlers must support and respect the Robots Exclusion Protocol.
2. Crawlers must be easily identifiable through their user agent string.
3. Crawlers must not interfere with the normal operation of a site.
4. Crawlers must support caching directives.

> 5. Crawlers must expose the IP ranges they are crawling from in a standardized format.

Author: No, I wasn't aware of this? Who is we? I would guess that this format is a google invention but you should know better than me...

> 5. Crawlers must expose the IP ranges they are crawling from in a standardized format.

Reviewer: I'm not actually sure where it's coming from to be honest. I borrowed it from another Google IP dump indeed, but it was already used by two other crawler operators at that point. I'll poke a bit more around to see if it has a name and if it's not a standard format, maybe we need someone from a networks WG (RTG? NMRG?) to help make it a standard

> 5. Crawlers must expose the IP ranges they are crawling from in a standardized format.

Reviewer: Done: https://github.com/garyillyes/cbcp/issues/4

6. Crawlers must expose a page where they explain how the crawled data is used and how it can be blocked.

## Crawlers must respect the Robots Exclusion Protocol

Crawlers must support the REP as defined in
[RFC9309](https://www.rfc-editor.org/rfc/rfc9309.html#section-2.2.1) to allow site owners
to opt out from crawling. Further crawler must respect the restrictions provided

> by the REP such as crawling delay and rate limits.

Reviewer: To the best of my knowledge all crawlers ignore "Crawl-delay", with only Bingbot parsing it out (but not actually using it?). I'll close this but opened an issue to [add information on self limiting](https://github.com/garyillyes/cbcp/issues/3#issue-3204596214)

> by the REP such as crawling delay and rate limits.

Author: Should we remove this sentence for now or keep it in for discussion?

> to opt out from crawling. Further crawler must respect the restrictions provided
> by the REP such as crawling delay and rate limits.

Author: ```suggestion
to opt out from crawling.
```

> by the REP such as crawling delay and rate limits.

Reviewer: Let's remove it. crawl-delay and other site owner provided rate limiting is highly unlikely to ever work in the favor of the site owner, no matter how transparent we go with how we crawl, and HTTP already has mechanisms to signal server issues (429, the suite of 500s)

> by the REP such as crawling delay and rate limits.

Reviewer: Resolving this for now so we have a clean slate

[RFC9110](https://www.rfc-editor.org/rfc/rfc9110#name-server-error-5xx), preferably also
additional heuristics such as relative response time of the server.

As such, crawlers should log already visited URL and how many requests they sent to a resource 
as well as the respective HTTP status codes in the responses,
especially if error occur, to avoid repeatedly crawling the same source.

> Generally, crawlers should avoid multiple concurrent requests to the same ressources

Author: ```suggestion
Generally, crawlers should avoid multiple concurrent requests to the same resources
```

> and use a crawling rate that is no higher than 1 request per second or as specified in

Reviewer: This is not going to work and I would be very surprised if *any* crawler operator would agree to this. 1 QPS to most websites, including news like WSJ and CNN, but also ecomm like Amazon, would mean spending about 95 years to crawl them. Sites like Facebook add more than 86400 unique urls a day.

Author: Okay, you are right this doesn't work for large pages. It's just a recommendation I read somewhere but didn't think too much about. Maybe we can add some text that discusses a bit more that rate should be adopted based on provider knowledge and for some infrastructures a lower rate should be implemented. Is that something that would be realistic?

Reviewer: Maybe we add something about this in [this issue](https://github.com/garyillyes/cbcp/issues/3)?

Author: Is this okay?
```suggestion
and should limit the crawling rate to avoid overload, if possible considering limits specified in
```

> the REP protocol. Further, ressources should not be re-crawled more often than every 24 hours.

Reviewer: This is not going to work. News sites' homepage is the hub that's hit about every 5 seconds because they care about having their fresh content in search and they publish often enough that having that 5s hit is critical.

In general any hard limit will not work I reckon, but happy to be proven wrong.

Author: Is this still useful to say?
```suggestion
the REP protocol. Further, ressources should not be re-crawled too frequently.
```

Ideally crawlers should limit the crawling depth and number of requests
per ressource to avoid loops.

> Crwalers should not attempt to access restricted ressources that requires a login, used CAPTCHAs, or are behind paywalls.

Reviewer: We specifically work with publishers to crawl behind paywalls: https://developers.google.com/search/docs/appearance/structured-data/paywalled-content

Author: Okay, we can add a clause like "if not explicitly agreed with the publishers". Would that work?

Or maybe we can even change "should not attempt to access" to "should not try to circumvent" or something? Would that make sense?

Reviewer: Yeah, perhaps something like "should not try to circumvent authentication"?

Author: Is this okay?
```suggestion
Crawlers should not try to circumvent authentication or other access restrictions e.g. when a login is required, CAPTCHAs are used, or the content is behind a paywall, except if explicitly negotiated with the website owner.
```

> Crawlers should only access content and by all means avoid to modify content by using
> HTTP request other than GET such as POST, PUT, DELETE. Similarly crawlers should avoid
> assessing Javascript content. 

Author: How about this?
```suggestion
Crawlers should access resources primarily with HTTP GET requests, resorting to other methods (e.g., POST, PUT) only if there's preexisting agreement with the publisher or the publisher's content management software makes those calls automatically when JavaScript is executed. Generally, load caused by executing JavaScript should be
considered carefully or even be avoided whenever possible.
```

> assessing Javascript content. 

Reviewer: JS content is imperative for a decent index nowadays. This is not going to work.
https://developers.google.com/search/docs/crawling-indexing/javascript/javascript-seo-basics

Author: Yes, I thought that was to restrictive but I thought I wait what you say... Are you crawling all JS content or do you select somehow?
